apps/settings/test/unit/modules/bluetooth/bluetooth_connection_manager_test.js

```diff
@@ -0,0 +1,552 @@
+      });
+
+      test('should clean up instance of cache while promise reject ',
+           function(done) {
+        connectionManager.getConnectedDevices().then(() => {
+          assert.ok(connectionManager.getConnectedDevices);
+          assert.isNull(connectionManager._getConnectedDevicesPromise);
+        }, () => {
+          // reject case
+        }).then(done, done);
+      });
+    });
+  });
+
+  suite('_resetConnectionInfo > ', function() {
+    setup(function() {
+      connectionManager.connectingAddress = 'AA:BB:CC:00:11:22';
+      connectionManager._getConnectedDevicesPromise = Promise.resolve();
+    });
+
+    test('connecting address, instance of cache should be reset ', function() {
+      connectionManager._resetConnectionInfo();
+      assert.isNull(connectionManager.connectingAddress);
+      assert.isNull(connectionManager._getConnectedDevicesPromise);
+    });
+  });
+
+  suite('_watchDefaultAdapterOnattributechanged > ', function() {
+    var mockAdapter;
+    setup(function() {
+      mockAdapter = {
+        addEventListener: function() {}
+      };
+      this.sinon.stub(mockAdapter, 'addEventListener');
+      this.sinon.stub(connectionManager, '_onAdapterAttributeChanged');
+    });
+
+    test('attributechanged should be regedit callback function ', function() {
+      connectionManager._watchDefaultAdapterOnattributechanged(mockAdapter);
+      assert.equal(mockAdapter.addEventListener.args[0][0], 'attributechanged');
+      mockAdapter.addEventListener.args[0][1]();
+      assert.isTrue(connectionManager._onAdapterAttributeChanged.calledWith(
+        mockAdapter));
+    });
+  });
+
+  suite('_unwatchDefaultAdapterOnattributechanged > ', function() {
+    var mockAdapter;
+    setup(function() {
+      mockAdapter = {
+        removeEventListener: function() {}
+      };
+      this.sinon.stub(mockAdapter, 'removeEventListener');
+      this.sinon.stub(connectionManager, '_onAdapterAttributeChanged');
+    });
+
+    test('onattributechanged should be unregedit ', function() {
+      connectionManager._unwatchDefaultAdapterOnattributechanged(mockAdapter);
+      assert.equal(mockAdapter.removeEventListener.args[0][0], 
+        'attributechanged');
+      assert.equal(mockAdapter.removeEventListener.args[0][1], 
+        connectionManager._onAdapterAttributeChanged);
+    });
+  });
+
+  suite('_onAdapterAttributeChanged > ', function() {
+    var mockAdapter, evt;
+    setup(function() {
+      mockAdapter = {
+        state: 'enabled'
+      };
+    });
+
+    suite('state changed in enabled ', function() {
+      setup(function() {
+        evt = {
+          attrs: ['state']
+        };
+        this.sinon.stub(connectionManager, '_restoreConnection');
+      });
+
+      test('_restoreConnection() should be called ', function() {
+        connectionManager._onAdapterAttributeChanged(mockAdapter, evt);
+        assert.isTrue(connectionManager._restoreConnection.called);
+      });
+    });
```

Comment:
This function is important. We should ensure `_restoreConnection` is not called in other states.
```
['enabling', 'disabling', 'disabled'].forEach(function(state) {
  test('_restoreConnection() should not be called when ' + state, function() {
    mockAdapter.state = state;
    connectionManager._onAdapterAttributeChanged(mockAdapter, evt);
    assert.isTrue(connectionManager._restoreConnection.notCalled);
  });
});
```